src/libstd/fs.rs

```diff
@@ -2261,11 +2282,41 @@ mod tests {
     }
 
     #[test]
+    fn concurrent_recursive_mkdir() {
+        for _ in 0..50 {
+            let mut dir = tmpdir().join("a");
```

Comment:
I think the main problem here is this use of `tmpdir()`. It deletes the path it returns when it goes out of scope and because it's not assigned to a variable that's immediately. See [here](https://github.com/rust-lang/rust/blob/4853584000c2e71710318b1ff3393b6be25c076a/src/libstd/sys_common/io.rs#L63-L92). Changing this to the following should fix that:

```rust
let tmpdir = tmpdir();
let mut dir = tmpdir.join("a");
```